conda_forge_tick/update_recipe/v2/version.py

```diff
@@ -0,0 +1,135 @@
+from __future__ import annotations
+
+import copy
+import logging
+import re
+from typing import TYPE_CHECKING, Literal
+
+from conda_forge_tick.update_recipe.v2.context import load_recipe_context
+from conda_forge_tick.update_recipe.v2.jinja import jinja_env
+from conda_forge_tick.update_recipe.v2.source import Source, get_all_sources
+from conda_forge_tick.update_recipe.v2.yaml import _dump_yaml_to_str, _load_yaml
+from conda_forge_tick.update_recipe.version import _try_url_and_hash_it
+
+if TYPE_CHECKING:
+    from pathlib import Path
+
+logger = logging.getLogger(__name__)
+
+HashType = Literal["md5", "sha256"]
+
+
+class CouldNotUpdateVersionError(Exception):
+    NO_CONTEXT = "Could not find context in recipe"
+    NO_VERSION = "Could not find version in recipe context"
+
+    def __init__(self, message: str = "Could not update version") -> None:
+        self.message = message
+        super().__init__(self.message)
+
+
+class HashError(Exception):
+    def __init__(self, url: str) -> None:
+        self.url = url
+        self.message = f"Could not hash {url}"
+        super().__init__(self.message)
+
+
+class Hash:
+    def __init__(self, hash_type: HashType, hash_value: str) -> None:
+        self.hash_type = hash_type
+        self.hash_value = hash_value
+
+    def __str__(self) -> str:
+        return f"{self.hash_type}: {self.hash_value}"
+
+
+def _has_jinja_version(url: str) -> bool:
+    """Check if the URL has a jinja `${{ version }}` in it."""
+    pattern = r"\${{\s*version"
+    return re.search(pattern, url) is not None
+
+
+def update_hash(source: Source, url: str, hash_: Hash | None) -> None:
```

Comment:
Some people will set the hash in the context section. yay edge cases!

Reply:
Yeah it's pretty discouraged by the schema right now and I don't see the benefit for the users to set the hash in the context. We just inject it in the source and the old hash will persist in the context for now. 
If someone gets annoyed we can modify this.

Reply:
This could break some recipes and people will be even more annoyed by that. We need to fix it.

Reply:
I made a lint to prevent this in conda-smithy: https://github.com/conda-forge/conda-smithy/pull/2028

Still needs tests etc. but I think that's the better way forward.

Reply:
@beckermr how would it break a recipe though. You would have:

```
context:
  sha256: foo

source:
  url: bla
  sha256: "1234..."
```

At some point someone would come and clean up the left-over hash from the context.

It would only break if the SHA hash was part of the URL. But in that case, the autotick updater has basically no chance anyways :)